### DISCUSSION TITLE
MockMvc와 Restassued의 차이점을 말해주세요.

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#208] [Spring Test]
뭘까용

**Comment by nayonsoso** [comment]
### MockMvc
- 스프링에서 제공하는 테스트 라이브러리의 일종
- 스프링 컨테이너를 띄우지 않고 MVC의 동작을 재현할 수 있다.
- 특히 컨트롤러의 요청 매핑, 필터, 스프링 시큐리티 처리 등을 테스트할 수 있다.
- MockMvc는 단독으로 사용되기도 하며 (단독으로 사용하기 위해선 MockMvcBuilder 를 사용한다 [참고](https://donghyeon.dev/spring/2019/03/28/Spring-컨트롤러-테스트를-위한-MockMvc/)), 스프링의 @WebMvcTest과 같은 다른 테스트 설정과 함께 사용될 수 있다.

### RestAssured
- 전 구간을 테스트하는 end-to-end 테스트에 초점을 둔 테스트 도구
- 실제 요청을 보내기 위해 @SpringBootTest 어노테이션을 함께 사용해야 한다.
  - 이때, 어플리케이션 실행과 테스트가 다른 포트에서 이뤄지게 하고 싶어 `WebEnvironment.RANDOM_PORT ` 를 사용한다면 RestAssured 도 동일한 포트를 사용하도록 설정해줘야 한다.
  - 왜냐하면 RestAssured는 디폴트로 localhost:8080 으로 요청을 보내기 때문이다.
  - By default REST assured assumes host localhost and port 8080 when doing a request. (https://github.com/rest-assured/rest-assured/wiki/Usage)

---

### 속도
당연히 필요한 빈만 불러오는 MockMvc가 더 빠르다.

### 의존성
MockMvc 는 SpringTest 에서 제공하는 라이브러리이기 때문에 spring-test 의존성이 추가되어있다면, 별도의 의존성을 추가하지 않아도 된다. 반면 RestAssured 는 io.rest-assured 의존성을 추가해야 한다. 의존성 추가가 별거냐고 할 수도 있지만, 의존성은 추가할 수록 프로젝트가 무거워지기 때문에 신중히 생각해야 한다.

### 가독성
MockMvc 는 perform, accept, andExpect 를 사용하는 반면, RestAssured 는 BDD 스타일로 given, when, then을 사용한다. 그리고 ResponseBody의 JSON을 검증하는 부분에서 MockMvc는 jsonpath 를 사용하고, RestAssured는 extract().as() 로 객체를 꺼내올 수 있다.
따라서 RestAssured 가 더 가독성이 좋다고 할 수 있다.

---

#### 개인적으로 MockMvc 불편한 점
프레젠테이션 레이어에 관련된 빈을 사용하기 위해, 이들이 의존하는 빈을 목빈으로 등록해줘야 에러가 안난다. 예를 들면, ThemeController 를 테스트하기 위해 컨트롤러가 직접 의존하는 ThemeService를 추가해줘야 하는 것은 물론이거니와,  ArgumentResolver 가 의존하는 MemberService와 JwtProvider까지 모킹해줘야 한다. 그런데 그것들 하나하나 모킹하는게 너무 헷갈림💀 그리고 그것들이 테스트 코드 윗부분에 쭉~ 적혀있으면 내가 컨트롤러 클래스만 단위테스트를 하고 있다는게 와닿지가 않는달까..

ref. https://tecoble.techcourse.co.kr/post/2020-08-19-rest-assured-vs-mock-mvc/

**Comment by gyungchan-jo** [comment]
### MockMvc

MockMvc는 Spring MVC 애플리케이션을 테스트하기 위해 제공되는 라이브러리입니다. 대부분 Controller Layer Unit Test(단위 테스트)에 사용됩니다. 보통`@WebMvcTest`를 통해 Presentation Layer Bean들만 불러온 후, 그 외 Bean은 Mock 객체 설정을 해주어 순수한 Controller 로직을 테스트합니다. 실제 서버를 시작하지 않고도 Spring MVC의 동작을 테스트할 수 있습니다.

### RestAssured

RestAssured는 RESTful 웹 서비스의 엔드포인트를 테스트하기 위한 라이브러리입니다. `@SpringBootTest`로 실제 요청을 보내서 전체적인 로직을 테스트하고, 실제 요청 시 필요한 다양한 메서드도 제공합니다.

---

차이점

- 의존성
MockMvc는 Spring Framework Test 클래스 중 하나다. 즉 Spring test 의존성이 추가되어있는 경우 별도의 의존성 추가를 하지 않아도 사용할 수 있다. 반면에 RestAssured는 직접 의존성을 추가해줘야 한다.

- 속도
RestAssured는 `@SpringBootTest`로 수행하고, MockMvc는 보통 `@WebMvcTest`로 수행한다.
`@SpringBootTest`로 테스트를 수행하면 등록된 Spring Bean을 전부 로드하기 때문에 시간이 오래 걸린다. 반면에 `@WebMvcTest`는 Presentation Layer의 Bean들만 로드하기 때문에 시간이 상대적으로 빠르다.

- 가독성 
RestAssured는 BDD(given, when, then) 스타일로 작성할 수 있어 가독성이 좋다.

[참고 블로그](https://tecoble.techcourse.co.kr/post/2020-08-19-rest-assured-vs-mock-mvc/)
